### PR TITLE
Enhancement/6256 key metrics settings endpoint

### PR DESCRIPTION
**Pull request by eugene-manuilov**
## Summary

Addresses issue:

- #6256 

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [ ] I have added a QA Brief on the issue linked above.
- [ ] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: includes/Core/Key_Metrics/Key_Metrics_Settings.php [hunk 1]
```diff
@@ -49,6 +49,28 @@ protected function get_default() {
...
+	public function merge( array $partial ) {
+		$settings = $this->get();
+
+		if ( isset( $partial['isWidgetHidden'] ) && is_bool( $partial['isWidgetHidden'] ) ) {
+			$settings['isWidgetHidden'] = $partial['isWidgetHidden'];
+		}
+
+		if ( ! empty( $partial['widgetSlugs'] ) && is_array( $partial['widgetSlugs'] ) ) {
+			$settings['widgetSlugs'] = $partial['widgetSlugs'];
+		}
```
Comment (techanvil) [inline includes/Core/Key_Metrics/Key_Metrics_Settings.php:69]: Wouldn't it be better to add `is_bool` and `is_array` checks to the conditions in `get_sanitize_callback`, rather than here?

File: includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+			return current_user_can( Permissions::VIEW_SPLASH ) || current_user_can( Permissions::VIEW_DASHBOARD );
+		};
+
+		return array(
+			new REST_Route(
+				'core/user/data/key-metrics',
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => function () {
+						return new WP_REST_Response( $this->settings->get() );
```
Comment (techanvil) [inline includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php:93]: Your implementation is fine but I notice the IB specifies using `rest_ensure_response` to create the REST responses, just want to flag this up.

File: includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+					},
+					'permission_callback' => $has_capabilities,
+					'args'                => array(
+						'data' => array(
+							'type'       => 'object',
+							'required'   => true,
+							'properties' => array(
+								'settings' => array(
+									'type'     => 'object',
+									'required' => true,
```
Comment (techanvil) [inline includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php:124]: We may as well include the `properties` for `settings` here too, what do you think? I.e.

```php
									'properties' => array(
										'isWidgetHidden' => array(
											'type'     => 'boolean',
											'required' => true,
										),
										'widgetSlugs'    => array(
											'type'     => 'array',
											'required' => true,
											'items'    => array(
												'type' => 'string',
											),
										),
```

**Review by techanvil — CHANGES_REQUESTED** [review]
This is looking good @eugene-manuilov. I've left a few minor comments.

In addition, I think it would be useful to add a bit of API testing to the QA Brief, along similar lines to what is specified in the QAB for #6172, i.e testing with `API.get` / `API.set` in the JS console.

```js
// This sort of thing:
googlesitekit.api.set('core', 'user', 'key-metrics', { settings: { isWidgetHidden: true, widgetSlugs:  ['analytics']} });
 
googlesitekit.api.get('core', 'user', 'key-metrics');
```

File: tests/phpunit/integration/Core/Key_Metrics/REST_Key_Metrics_ControllerTest.php [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+		$this->settings->register();
+		$this->settings->set( $original_settings );
+
+		$request  = new WP_REST_Request( 'GET', '/' . REST_Routes::REST_ROOT . '/core/user/data/key-metrics' );
+		$response = rest_get_server()->dispatch( $request );
+
+		$this->assertEqualSetsWithIndex( $original_settings, $response->get_data() );
+	}
+
+	public function test_set_settings() {
```
Comment (techanvil) [inline tests/phpunit/integration/Core/Key_Metrics/REST_Key_Metrics_ControllerTest.php:88]: It might also be worth adding a test for the error handling behaviour...

File: includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php (hunk 2 above)
Comment (eugene-manuilov) [reply includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php:93]: The `rest_ensure_response` function isn't needed because it just adds an overhead and internally creates a new instance of the rest response like I did:

https://github.com/WordPress/wordpress-develop/blob/6.1/src/wp-includes/rest-api.php#L637

File: includes/Core/Key_Metrics/Key_Metrics_Settings.php (hunk 1 above)
Comment (eugene-manuilov) [reply includes/Core/Key_Metrics/Key_Metrics_Settings.php:69]: It will be too late. The idea here is that if some properties don't match expected type, then we don't merge it into settings.

File: tests/phpunit/integration/Core/Key_Metrics/REST_Key_Metrics_ControllerTest.php (hunk 4 above)
Comment (eugene-manuilov) [reply tests/phpunit/integration/Core/Key_Metrics/REST_Key_Metrics_ControllerTest.php:88]: Yep, added.

File: includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php (hunk 3 above)
Comment (eugene-manuilov) [reply includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php]: Sure, added. Thanks!

Comment (eugene-manuilov) [reply includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php]: Added them as not required because we allow some properties to be missed.

File: includes/Core/Key_Metrics/Key_Metrics_Settings.php (hunk 1 above)
Comment (techanvil) [reply includes/Core/Key_Metrics/Key_Metrics_Settings.php:69]: Hmm, OK. It seems a bit odd though, to have a check in `merge` to ensure the exact type is passed in, but then have code in `get_sanitize_callback`, to massage the data into the correct format if the exact type is _not_ passed in. Why not just pass the data straight through and let `get_sanitize_callback` take care of it? It will ensure that no incorrectly formed data will be stored and it would be easier to manage keeping all data checks of this sort within the one method...

**Review by techanvil — CHANGES_REQUESTED** [review]
Thanks @eugene-manuilov. I've left another comment on the `merge` thread, other than that, looks good to go.

File: includes/Core/Key_Metrics/Key_Metrics_Settings.php (hunk 1 above)
Comment (eugene-manuilov) [reply includes/Core/Key_Metrics/Key_Metrics_Settings.php:69]: @techanvil, ok let's assume I did what you asked and moved sanitization to the `get_sanitize_callback` callback. Let's see how it works now. Assuming we have the following settings:
```php
array(
    'widgetSlugs'    => array( 'moduleA', 'moduleB' ),
    'isWidgetHidden' => false,
)
```
Then the `merge` function is called with wrong data:
```php
$key_metrics_settings->merge( array(
    'widgetSlugs'    => null,
    'isWidgetHidden' => true,
) );
```
Since we don't validate the incoming data in the `merge` function, we have wrong data merged and after the sanitization callback is executed, we will get the `widgetSlugs` array reset to an empty array because `null` is invalid value - as a result we get data loss.

In other words instead:
```php
array(
    'widgetSlugs'    => array( 'moduleA', 'moduleB' ),
    'isWidgetHidden' => true,
)
```
We will get:
```php
array(
    'widgetSlugs'    => array(),
    'isWidgetHidden' => false,
)
```

> It seems a bit odd though, to have a check in `merge` ...

There is nothing odd there. The `get_sanitize_callback` callback is responsible for data sanitization before saving it into the database. The sanitization in the `merge` function is responsible for the incoming data verification before changing the current settings.

Comment (eugene-manuilov) [reply includes/Core/Key_Metrics/Key_Metrics_Settings.php:69]: > However, it raises the question, why should `merge` behave differently to `set`?

Because it allows partial updates.

> If we move the check for `is_array` etc into `get_sanitize_callback` then we can have consistent behaviour for both `merge` and `set`

The `get_sanitize_callback` callback already takes care of the `widgetsSlug` property by calling the `Sanitize:: sanitize_string_list` method. I don't think we should change anything there. At least not in this ticket, because it is out of scope for this one.

https://github.com/google/site-kit-wp/blob/8825f41e392caa787480ae033c8eed5d0e4e174a/includes/Core/Util/Sanitize.php#L30-L47

Comment (techanvil) [reply includes/Core/Key_Metrics/Key_Metrics_Settings.php:69]: What I meant was, why should the validation/sanitization for `merge` behave differently to `set`? It just seems a bit odd to be have, say, `merge( array( 'widgetSlugs' => 'foo' ) )` result in the settings _not_ being updated, while the same arguments to `set` results in the data being updated (with `foo` wrapped in an array).

Conceptually I think `merge` should just be like `set` except it only sets a slice of the data.

Sorry, I don't mean to be obtuse or excessively pedantic here, it's a pretty small detail but I do think it's a good opportunity to avoid introducing an inconsistent interface before it's actually used anywhere. I do take your point that it's arguably out of scope to modify `get_sanitize_callback`, but in that case why not just pass the data through without checking the type? It will still be sanitized by `get_sanitize_callback` which can be our single source of truth for data sanitization.

Also as you've mentioned elsewhere, by adding the API schema we have an extra layer of validation that will enforce these data types for the API calls.

Comment (eugene-manuilov) [reply includes/Core/Key_Metrics/Key_Metrics_Settings.php]: > but in that case why not just pass the data through without checking the type? 

Sounds good, I have reverted my changes and copied the `merge` method from the `Module_Settings` class as IB suggests.

File: includes/Core/Key_Metrics/Key_Metrics_Settings.php [hunk 5]
```diff
@@ -49,6 +49,27 @@ protected function get_default() {
...
+	 */
+	public function merge( array $partial ) {
+		$settings = $this->get();
+		$partial  = array_filter(
+			$partial,
+			function ( $value ) {
+				return null !== $value;
+			}
+		);
+		$updated  = array_intersect_key( $partial, $settings );
```
Comment (techanvil) [inline includes/Core/Key_Metrics/Key_Metrics_Settings.php:68]: @eugene-manuilov thanks for updating this method, however, I notice when testing the endpoint (and as the JSDoc does mention) that only existing keys will be updated, so when trying to POST when the settings have the default value of an empty array, the update is ignored.

My understanding is that the settings will only be updated in response to user input, which means we need to be able to merge them from the default value. With that in mind, I think we need a slight variation on this implementation. Either by specifying the list of allowed keys here instead of using `$settings` to determine the allowed keys, or potentially by updating `get_sanitize_callback` to drop unknown fields.

This also implies there should be an additional test case to test merging when the settings are not yet present.

File: includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+									'required'   => true,
+									'properties' => array(
+										'isWidgetHidden' => array(
+											'type'     => 'boolean',
+											'required' => false,
+										),
+										'widgetSlugs'    => array(
+											'type'     => 'array',
+											'required' => false,
+											'items'    => array(
```
Comment (techanvil) [inline includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php:129]: I've spotted something that was missed in the AC. From the [Design Doc](https://docs.google.com/document/d/1k8NOo2xEHUQAcgYG0TB5sUjGcPiC8SWIalEdd9G8-9g/edit#heading=h.th9qnr6nkrac):

> The `POST:key-metrics` endpoint should also return an error (which should be displayed to the user) if more than four metrics were submitted somehow.

It doesn't appear to have been captured in any of the other issues. This can easily be accomplished using [`maxItems`](https://json-schema.org/understanding-json-schema/reference/array.html#id12) here:

```suggestion
											'maxItems' => 4,
											'items'    => array(
```

It would also be useful to cover an another test.

I realise this isn't in the AC so if you feel it's scope creep we can always create a separate issue for it, but I think it would be nice to include here.

**Review by techanvil — CHANGES_REQUESTED** [review]
Thanks @eugene-manuilov. I've noticed a couple more things, please take a look and see what you think.

File: includes/Core/Key_Metrics/Key_Metrics_Settings.php [hunk 7]
```diff
@@ -49,6 +49,37 @@ protected function get_default() {
...
+		}
+
+		$partial = array_filter(
+			$partial,
+			function ( $value ) {
+				return null !== $value;
+			}
+		);
+
+		$updated = array_intersect_key( $partial, $settings );
```
Comment (techanvil) [inline includes/Core/Key_Metrics/Key_Metrics_Settings.php:78]: Hi @eugene-manuilov, thanks for making changes here, however IMHO it's still not quite right.

Due to setting these default values, if `merge` is called without `isWidgetHidden`, the result of calling `get_sanitize_callback` on the empty string value will end up being set, in other words `isWidgetHidden` will default to `false`.

It's also a bit unclear to set default values which are not the expected data type of the settings. If we _do_ want to default `isWidgetHidden` then I'd make it explicit here by setting it to `false` instead of the empty string. However, I'd suggest that instead, we don't set defaults and take an approach like the following:

```suggestion
		$partial = array_filter(
			$partial,
			function ( $value ) {
				return null !== $value;
			}
		);

		$allowed_settings = array(
			'widgetSlugs'    => true,
			'isWidgetHidden' => true,
		);

		$updated = array_intersect_key( $partial, $allowed_settings );
```

**Review by techanvil — CHANGES_REQUESTED** [review]
Hi @eugene-manuilov, please see my comment on `merge`.

Additionally, for some reason the PHP 5.6 unit tests are failing in CI, I have tried re-running but got the same result. Please can you take a look?

File: includes/Core/Key_Metrics/Key_Metrics_Settings.php [hunk 8]
```diff
@@ -49,6 +49,37 @@ protected function get_default() {
 		return array();
 	}
 
+	/**
+	 * Merges an array of settings to update. Only existing keys will be updated.
```
Comment (techanvil) [inline includes/Core/Key_Metrics/Key_Metrics_Settings.php:53]: This JSDoc should also be updated:
```suggestion
	 * Merges an array of settings to update.
```

File: includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+					'callback'            => function ( WP_REST_Request $request ) {
+						// Data is already validated because we've defined the detailed schema.
+						// If the incoming data param doesn't match the schema, then WordPress
+						// will automatically return the rest_invalid_param error and we will
+						// never get to here.
+						$data     = $request->get_param( 'data' );
+						$settings = $data['settings'];
+
+						// Additional check is needed to ensure that we have no more than 4 widget
+						// slugs provided.
```
Comment (techanvil) [inline includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php:111]: Adding a bit of context to the comment as discussed.

```suggestion
						// slugs provided. This is required until we drop support for WP versions below 5.5.0, after
						// which we can solely rely on `maxItems` in the schema validation (see below).
						// See https://github.com/WordPress/WordPress/blob/965fcddcf68cf4fd122ae24b992e242dfea1d773/wp-includes/rest-api.php#L1922-L1925.
```

**Review by techanvil — APPROVED** [review]
LGTM, nice one @eugene-manuilov!

File: includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+						// will automatically return the rest_invalid_param error and we will
+						// never get to here.
+						$data     = $request->get_param( 'data' );
+						$settings = $data['settings'];
+
+						// Additional check is needed to ensure that we have no more than 4 widget
+						// slugs provided. This is required until we drop support for WP versions below 5.5.0, after
+						// which we can solely rely on `maxItems` in the schema validation (see below).
+						// See https://github.com/WordPress/WordPress/blob/965fcddcf68cf4fd122ae24b992e242dfea1d773/wp-includes/rest-api.php#L1922-L1925.
+						if ( count( $settings['widgetSlugs'] ) > 4 ) {
```
[the remark this answers is not in the document]
Comment (techanvil) [reply includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php:114]: That's a good spot, @aaemnnosttv.

In hindsight, looking at the AC, I think that in fact we should make both `widgetSlugs` and `isWidgetHidden` mandatory, as the AC describes the request shape as follows, without indicating they should be optional.

> - the data request should be of the following shape:
```js
  {
      widgetSlugs,    // Object of widget slug strings,
      isWidgetHidden    // boolean value
  }
```

@eugene-manuilov, would you mind creating a followup PR to address this?
